### PR TITLE
Updates to acrobot parameter estimation with nonlinear estimator

### PR DESCRIPTION
**Pull request by hanssusilo**
-The lines with oldpath seem deprecated - took that out
-Two acrobot systems are compared here, one which has the nominal parameter values, and one with the true parameter values, which is randomly distributed but centered at the nominal values.  The true system is used to generate the trajectory and observations.  Of course, the true system is unknown to the parameterEstimation script, but the nominal system is passed into parameterEstimation and is used to extract the system's polynomial representation and seed the nonlinear estimator
-Options are set at the top of the paramEstSyntheticData file, which allows user to play with different noise settings and methods of obtaining acceleration measurements as well as noise parameters
-Rewrote parameEstimation to utilize the functions Andres made and clean up code
-generation of all observable measurements taken out from parameterEstimation and given to paramEstSyntheticData, as parameterEstimation should really just be given any data, indifferent to the exact nuances of the system.  Choice of what data to pass into parameterEstimation should be left to the user, so that they have control over sensor processing/filtering.
-Took out geometric programming portion - don't worry, it's saved in my branch for later use.

### CODE REVIEW

File: drake/examples/Acrobot/test/paramEstSyntheticData.m [hunk 1]
```diff
@@ -1,26 +1,99 @@
 function paramEstSyntheticData
 
-oldpath = addpath(fullfile(pwd,'..'));
```
Comment (RussTedrake) [inline drake/examples/Acrobot/test/paramEstSyntheticData.m]: i think you'll need this, although it's true we now use `addpathTemporary` instead.  See other test directory files.

File: drake/examples/Acrobot/test/paramEstSyntheticData.m [hunk 2]
```diff
@@ -1,26 +1,99 @@
...
-% Try parameter estimation without any noise
-Ts = .01; breaks=getBreaks(utraj); T0 = breaks(1); Tf = breaks(end);
-data = iddata(eval(xtraj,T0:Ts:Tf)',eval(utraj,T0:Ts:Tf)',Ts,'InputName',r.getInputFrame.getCoordinateNames(),'OutputName',r.getOutputFrame.getCoordinateNames());
+r = AcrobotPlant;
+rtrue = AcrobotPlant;
 
-estimated_parameters = parameterEstimation(r,data);
+if mode ~= 0
+    % Perturb original parameter estimates with random percentage error
+    % normally distributed with standard dev = paramstd, and greater than -1
```
Comment (RussTedrake) [inline drake/examples/Acrobot/test/paramEstSyntheticData.m]: why the greater than -1 bit?  are you trying to make sure that the parameters do not go negative?  wouldn' t it be better to enforce that directly below?

File: drake/systems/plants/@Manipulator/parameterEstimation.m [hunk 3]
```diff
@@ -25,20 +25,43 @@
...
+% 
+% @option print_result determines if the function will print the results
+% 0 = no print, 1 = print estimated and original, 2 = only print estimated
+% 
+% @option method determines which method to use for estimation
+% 0 = no print, 1 = print estimated and original, 2 = only print estimated
+% 
+
+%% handle options
+if (nargin>2 && isstruct(varargin{1})) options=varargin{1};
```
Comment (RussTedrake) [inline drake/systems/plants/@Manipulator/parameterEstimation.m]: if the only thing varargin can be is an options structure, then it should be named options and you set it manually if nargin < 4

File: drake/systems/plants/@Manipulator/parameterEstimation.m [hunk 4]
```diff
@@ -1,4 +1,4 @@
-function [phat,estimated_delay] = parameterEstimation(obj,data,options)
+function [phat,estimated_delay] = parameterEstimation(obj,obj2,data,varargin)
 %
```
Comment (RussTedrake) [inline drake/systems/plants/@Manipulator/parameterEstimation.m]: why two objects?  definitely need to document that interface change.

File: drake/systems/plants/@Manipulator/parameterEstimation.m [hunk 5]
```diff
@@ -25,20 +25,43 @@
...
+% 
+
+%% handle options
+if (nargin>2 && isstruct(varargin{1})) options=varargin{1};
+else options=struct(); end
+if (~isfield(options,'print_result')) 
+  options.print_result=0; 
+end
+if (~isfield(options,'method')) 
+  options.method=0;
```
Comment (RussTedrake) [inline drake/systems/plants/@Manipulator/parameterEstimation.m]: use string names instead of integer ids?

File: drake/systems/plants/@Manipulator/parameterEstimation.m [hunk 6]
```diff
@@ -118,75 +111,39 @@
...
-%  not hard, just not implmented yet.
-qpopt = optimset('Display','off');
-[log_p_est,~,exitflag] = quadprog(eye(np),-log(p_orig),[],[],A,log(lp_est),log(pmin),log(pmax),log(p_orig),qpopt);
 
-if (exitflag~=1)
-  warning('quadprog exitflag = %d',exitflag);
-end
+% Nonlinear least-squares solver
+nonlinfun = @(x) M_data*(msubs(lp,p,x))+Mb_data;
+[x, sqerr_est] = lsqnonlin(nonlinfun,p_orig,pmin,pmax);
```
Comment (RussTedrake) [inline drake/systems/plants/@Manipulator/parameterEstimation.m]: would be better to set up a NonlinearProgram object, so you can use snopt, etc.

File: drake/systems/plants/@Manipulator/parameterEstimation.m (hunk 5 above)
Comment (RussTedrake) [inline drake/systems/plants/@Manipulator/parameterEstimation.m]: you are still planning to add the method which does linear estimation of the identifiable parameters, then nonlinear least squares for the original params, right?

File: drake/examples/Acrobot/test/paramEstSyntheticData.m (hunk 2 above)
Comment (hanssusilo) [inline drake/examples/Acrobot/test/paramEstSyntheticData.m]: Yep, to keep it from going below zero.  I just wanted a truncated gaussian without peaks at the boundary values.  But I can enforce it with a value clip function if you think that's better.

File: drake/systems/plants/@Manipulator/parameterEstimation.m (hunk 4 above)
Comment (hanssusilo) [inline drake/systems/plants/@Manipulator/parameterEstimation.m]: Whoops, that was a test - changed it back to original.

File: drake/examples/Acrobot/test/paramEstSyntheticData.m (hunk 1 above)
Comment (hanssusilo) [inline drake/examples/Acrobot/test/paramEstSyntheticData.m]: Ok, I will add this line:
tmp = addpathTemporary(fullfile(pwd,'..'));

Comment (hanssusilo) [inline drake/examples/Acrobot/test/paramEstSyntheticData.m]: I took it out because Andres and I found that the line created errors with adding spotless to the environment path.  Did you know anything about that and if this new method fixes it?

File: drake/systems/plants/@Manipulator/parameterEstimation.m [hunk 7]
```diff
@@ -90,103 +133,124 @@
...
+        prog=prog.addQuadraticCost(eye(np),p_orig,1:np);
+        lpconstraint_handle = @(x) lpconstraint_fun(x,lp,p);
+        lpconstraint = FunctionHandleConstraint(lp_est,lp_est,nlp,lpconstraint_handle);
+        prog=prog.addConstraint(lpconstraint,1:np);
+    end
+    [x,F,info] = prog.solve(p_orig);
+    if(info ~= 1)
+    	error('failed to solve the problem');
+    end
+elseif strcmp(options.method,'lsqnonlin')
```
Comment (RussTedrake) [inline drake/systems/plants/@Manipulator/parameterEstimation.m]: we should support this as a fallback/option in the NonlinearProgram (perhaps we already do).  you shouldn't have to special case it here.
